### PR TITLE
Freestanding mode

### PR DESCRIPTION
**Pull request by Eisenwave**
This PR changes how authentication of Wheatley works. All information, including the `freestanding` flag is read from an `auth.json` file, which makes it possible to use *some* features outside of TCCPP without having to touch code.

### CODE REVIEW

**Review by jeremy-rifkin — APPROVED** [review]

File: src/components/wiki.ts [hunk 1]
```diff
@@ -136,12 +141,15 @@ export class Wiki extends BotComponent {
                     title: "query",
                     description: "Query",
                     required: true,
-                    autocomplete: query =>
-                        Object.values(this.articles)
+                    autocomplete: query => {
```
Comment (jeremy-rifkin) [inline src/components/wiki.ts:144]: Can we revert this change

File: src/infra/guild-command-manager.ts [hunk 2]
```diff
@@ -21,17 +21,32 @@ export class GuildCommandManager {
...
+                M.log("Sending application commands");
+                await rest.put(
+                    Routes.applicationCommands(this.wheatley.id),
+                    { body: this.commands },
+                );
+                // Clear any previous guild commands
+                await rest.put(
+                    Routes.applicationGuildCommands(this.wheatley.id, this.wheatley.guildId),
+                    { body: [] },
+                );
```
Comment (jeremy-rifkin) [inline src/infra/guild-command-manager.ts:34]: This `put(Routes.applicationGuildCommands)` is no longer needed since we never set `applicationGuildCommands`, I think we can clean this section up a lot. It should be ok to put `Routes.applicationCommands` in freestanding mode.

File: src/utils.ts [hunk 3]
```diff
@@ -300,10 +300,10 @@ export function init_debugger(_client: Discord.Client) {
     client = _client;
 }
 
-export async function critical_error(...args: any[]) {
+export function critical_error(...args: any[]) {
     M.error(...args);
-    try {
-        if(await get_zelis()) {
+    get_zelis().then(zelis_found => {
```
Comment (jeremy-rifkin) [inline src/utils.ts:305]: I think this is ok

File: src/utils.ts [hunk 4]
```diff
@@ -314,9 +314,10 @@ export async function critical_error(...args: any[]) {
                     } catch { void(0); }
                 }
             }
-            zelis!.send(`Critical error occurred: ${strs.join(" ")}`);
+            zelis!.send(`Critical error occurred: ${strs.join(" ")}`)
+                .catch(() => void(0));
         }
-    } catch { void(0); }
+    })
```
Comment (jeremy-rifkin) [inline src/utils.ts:320]: Super minor: Can we add a semicolon here

File: src/wheatley.ts [hunk 5]
```diff
@@ -80,9 +82,16 @@ type text_command_map_target = {
     deletable: boolean;
 };
 
+type authentication = {
+    id: string;
+    guild: string;
+    token: string;
+    freestanding?: boolean;
+}
```
Comment (jeremy-rifkin) [inline src/wheatley.ts:90]: Super minor: Can we add a semicolon here

File: src/wheatley.ts [hunk 6]
```diff
@@ -117,15 +126,25 @@ export class Wheatley extends EventEmitter {
...
+    // Guild ID, falls back onto TCCPP if not provided in auth.json.
+    readonly guildId: string;
+    // True if freestanding mode is enabled. Defaults to false.
+    readonly freestanding: boolean;
 
     constructor(readonly client: Discord.Client, readonly database: DatabaseInterface) {
         super();
 
-        this.guild_command_manager = new GuildCommandManager(this);
+        const auth: authentication = JSON.parse(fs.readFileSync("auth.json", { encoding: "utf-8" }));
```
Comment (jeremy-rifkin) [inline src/wheatley.ts:139]: I think it'd be better to read this async in main then pass `auth`

File: src/wheatley.ts [hunk 7]
```diff
@@ -80,9 +82,16 @@ type text_command_map_target = {
     deletable: boolean;
 };
 
+type authentication = {
+    id: string;
```
Comment (jeremy-rifkin) [inline src/wheatley.ts:86]: Would it be possible to just get the client's id after ready rather than specifying it in the json

File: src/infra/guild-command-manager.ts [hunk 8]
```diff
@@ -21,17 +21,32 @@ export class GuildCommandManager {
...
+            }
+            else {
```
Comment (jeremy-rifkin) [inline src/infra/guild-command-manager.ts:36]: Minor, but to match the style of the rest of the codebase can we update this to be
```ts
if(...) {
    ...
} else {
    ...
}
```
(if the if-else exists after the above comment)

**Review by jeremy-rifkin — CHANGES_REQUESTED** [review]
(Oops, accidentally clicked the submit earlier.)

Thank you for the contribution! This will be helpful for future contributors.

I've made a number of comments, some minor style things as well as some bigger changes.

Overall structurally: We can do a lot better than just making note of freestanding or not freestanding in comments on classes. It'd be better to add a `static get freestanding(): boolean;` in `BotComponent`, having components override as needed. Then in Wheatley components don't have to be split between `nonFreestandingSetup` and the regular setup routine.

File: src/infra/guild-command-manager.ts (hunk 2 above)
Comment (Eisenwave) [reply src/infra/guild-command-manager.ts]: I've cleaned up the section, but would prefer to use `applicationGuildCommands` for freestanding. Afaik, application guild commands update faster, so it's more convenient for quick little tests that you would do in freestanding mode.

**Comment by Eisenwave** [comment]
> Overall structurally: We can do a lot better than just making note of freestanding or not freestanding in comments on classes. It'd be better to add a `static get freestanding(): boolean;` in `BotComponent`, having components override as needed. Then in Wheatley components don't have to be split between `nonFreestandingSetup` and the regular setup routine.

You're probably right; the reason why I left the comments was mostly so that I know what the components do, same for future contributors. You probably could do do it per component like that, although there are differences in the two setups beyond just that.

I feel like this is a good idea, but it's not necessary to make this change to components within this PR, but a future refactoring PR.
